pkg/safemath/safemath.go

```diff
@@ -0,0 +1,33 @@
+package safemath
+
+import "math/bits"
+
+// This is a constant-time version of the following function:
+//
+//	func SafeOffset(x uint64, y int16) (res uint64, isOverflow bool) {
+//		res = x + uint64(y)
+//		if y < 0 {
+//			isOverflow = res >= x
+//		} else {
+//			isOverflow = res < x
+//		}
+//		return
+//	}
+//
+// This shows better results because the final bytecode
+// doesn't contain any conditional jump instructions
+// making it easier for a processor to pipeline the function.
+func SafeOffset(x uint64, y int16) (res uint64, isOverflow bool) {
+	// I'll leave proving that this is correct as an exercise for the reader :)
+	res = x + uint64(y)
+	// Why does this work?
+	// Let's proceed by cases on MSB(x).
+	// If MSB(x) == 1 then we don't have any potential overflow issues
+	// if y < 0 (i.e. MSB(y) == 1), meaning that we can consider the case y >= 0.
```

Comment:
Nitpick: both comments on line `25` and `26` are related but feel disconnected, could you paraphrase or reorganize, so that is more clear that we only need to analize for overflows when `MSB(x) == 1` and `MSB(y) == 0`